graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +593,194 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_tagged_expression(expression, current_locations):
+    """Replace non-existent tag Expressons with True, and simplify the result."""
+    if isinstance(expression, BinaryComposition):
+        if isinstance(expression.left, ContextField):
+            context_field = expression.left
+        elif isinstance(expression.right, ContextField):
+            context_field = expression.right
+        elif expression.operator == u'||':
+            if expression.left == TrueLiteral or expression.right == TrueLiteral:
+                return TrueLiteral
+            else:
+                return expression
+        elif expression.operator == u'&&':
+            if expression.left == TrueLiteral and expression.right == TrueLiteral:
+                return TrueLiteral
+            if expression.left == TrueLiteral:
+                return expression.right
+            if expression.right == TrueLiteral:
+                return expression.left
+            else:
+                return expression
+        else:
+            return expression
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in current_locations:
+            return TrueLiteral
+        else:
+            return expression
+    elif isinstance(expression, TernaryConditional):
+        if expression.predicate == TrueLiteral:
+            return expression.if_true
+        else:
+            return expression
+    elif isinstance(expression, BetweenClause):
+        lower_bound = expression.lower_bound
+        upper_bound = expression.upper_bound
+        if isinstance(lower_bound, ContextField) or isinstance(upper_bound, ContextField):
+            raise AssertionError(u'Found BetweenClause with ContextFields as lower/upper bounds. '
+                                 u'THis should never happen: {}'.format(expression))
+    else:
+        return expression
+
+
+def _construct_update_tags_visitor_fn(current_locations):
+    """Return an Expression updater using the given `current_locations`."""
+    def visitor_fn(expression):
+        return _update_tagged_expression(expression, current_locations)
+    return visitor_fn
+
+
+def _get_current_locations_from_match_traversals(match_traversals):
+    """Return a set of locations present in the given match traversals."""
+    current_locations = set()
+
+    for traversal in match_traversals:
+        for step in traversal:
+            if step.as_block is not None:
+                location_name, _ = step.as_block.location.get_location_name()
+                current_locations.add(location_name)
+    return current_locations
+
+
+def _lower_filters_in_match_traversals(match_traversals, visitor_fn):
+    """Return new match traversals, replacing filters involving non-existent tags with True."""
```

Comment:
`replacing filters involving non-existent tags with True` is somewhat misleading -- if the only filter is one with a `ContextField`, it will not be replaced with `True` but eliminated altogether.

Please write a full docstring and explain these subtleties.